util/rateLimit.go

```diff
@@ -0,0 +1,321 @@
+			return nil
+		default:
+			return fmt.Errorf("could not return capacity to client reservation")
+		}
+	}
+	// before adding to the sharedCapacity, confirm it wouldn't overprovision the sharedCapacity
+	maxShared := erl.MaxCapacity - (numClients * erl.CapacityPerReservation)
+	if len(erl.sharedCapacity) >= maxShared {
+		return fmt.Errorf("capacity return would overprovision the sharedCapacity")
+	}
+	select {
+	case erl.sharedCapacity <- capacity{}:
+		return nil
+	default:
+		return fmt.Errorf("could not return capacity to sharedCapacity")
+	}
+}
+
+// ContainsReservationFor is a lock-safe check for a client having a reservation
+func (erl ElasticRateLimiter) ContainsReservationFor(c client) bool {
+	erl.clientMu.RLock()
+	defer erl.clientMu.RUnlock()
+	_, exists := erl.capacityByClient[c]
+	return exists
+}
+
+// OpenReservation creates an entry in the ElasticRateLimiter's reservedCapacity map,
+// and optimistically transfers capacity from the sharedCapacity to the reservedCapacity
+func (erl ElasticRateLimiter) OpenReservation(c client) error {
+	erl.clientMu.Lock()
+	defer erl.clientMu.Unlock()
+	if _, exists := erl.capacityByClient[c]; exists {
+		return fmt.Errorf("client already has a reservation")
+	}
+	// guard against overprovisioning, if there is less than a reservedCapacity amount left
+	remaining := erl.MaxCapacity - (erl.CapacityPerReservation * len(erl.capacityByClient))
+	if erl.CapacityPerReservation > remaining {
+		return fmt.Errorf("not enough capacity to reserve for client: %d remaining, %d requested", remaining, erl.CapacityPerReservation)
+	}
+	// make capacity for the provided client
+	erl.capacityByClient[c] = make(chan capacity, erl.CapacityPerReservation)
+	delete(erl.closedClients, c)
+
+	// start asynchronously filling the capacity
+	// by design, the ElasticRateLimiter will overprovision capacity for a newly connected host,
+	// but will resolve the discrepancy ASAP by consuming capacity from the shared capacity which it will not return
+	for i := 0; i < erl.CapacityPerReservation; i++ {
+		select {
+		case erl.capacityByClient[c] <- capacity{}:
+		default:
+		}
+	}
+	go func() {
+		// allow this thread to stubbornly wait and consume any available capacity,
+		// since it needs to repay the capacity loaned to the client
+		for i := 0; i < erl.CapacityPerReservation; i++ {
+			<-erl.sharedCapacity
+		}
+	}()
+	return nil
+}
+
+// CloseReservation will remove the client mapping to capacity channel,
+// will mark the client as "closed", and will drain the client's channel to the shared channel
+func (erl ElasticRateLimiter) CloseReservation(c client) error {
+	erl.clientMu.Lock()
+	clientCh, exists := erl.capacityByClient[c]
+	// guard clauses, and preventing the ElasticRateLimiter from draining its own sharedCapacity
+	if !exists ||
+		clientCh == nil ||
+		clientCh == erl.sharedCapacity {
+		return fmt.Errorf("client not registered for capacity")
+	}
+	delete(erl.capacityByClient, c)
+	erl.closedClients[c] = true
```

Comment:
i would have done this with a go routine that tried to drain `CapacityPerReservation` and moved them to the shared queue.